File: demos/pi/pi_ring/worker.c
#include <helios.h>
#include <stdio.h>
#include <posix.h>

double eval(int position, int number_workers, int intervals);

int main(void)
{ int 		position, number_workers, temp, intervals;
  double	sum, total;
  
 		/* get the worker's position in the pipeline */
  read( 0, (BYTE *) &position, sizeof(int));
  temp = position + 1;
  write(1, (BYTE *) &temp, sizeof(int));

		/* get the length of the pipeline */
  read( 0, (BYTE *) &number_workers, sizeof(int));
  write(1, (BYTE *) &number_workers, sizeof(int));
   
		/* get the number of intervals per worker */
  read( 0, (BYTE *) &intervals, sizeof(int));
  write(1, (BYTE *) &intervals, sizeof(int));
  
  sum = eval(position, number_workers, intervals);
  
  read( 0, (BYTE *) &total, sizeof(double));
  total = total + sum;
  write(1, (BYTE *) &total, sizeof(double));
  
  return(0);
}

double eval(int position, int number_workers, int intervals)
{ int		first, current, last;
  double	width, sum, tmp;
  
  sum		= 0.0;
  width		= 1.0 / (double) (number_workers * intervals);
  first		= position * intervals;
  last		= first + intervals;
  
  for (current = first; current < last; current++)
   { tmp = (0.5 + (double) current) * width;
     sum = sum + width * (4.0 / (1.0 + tmp * tmp));
   }
   
  return(sum);
}      

File: demos/pi/pi_pasc/worker.pas
PROGRAM worker(input, out);

FUNCTION _read(hand:integer; place:integer; amount:integer):integer;EXTERNAL;
FUNCTION _write(hand:integer;place:integer;amount:integer):integer;EXTERNAL;

	{ the evaluation routine }
FUNCTION eval(position, workers, intervals : integer):longreal;
VAR
	first, current, last : integer;
	width, sum, tmp : longreal;

BEGIN
	sum	:= 0.0D0;
	width	:= 1.0D0 / (workers * intervals);
	first	:= position * intervals;
	last	:= first + intervals;

	for current := first to (last - 1) do
	BEGIN
		tmp := (0.5D0 + current) * width;
		sum := sum + width * (4.0D0 / (1.0D0 + tmp * tmp));
	END;

	eval := sum;
END;

	{ the main routine }
VAR	position, number_workers, intervals, junk, temp:integer;
	total, sum:longreal;

BEGIN
	junk := _read( 0, addr(position), 4);
	temp := position + 1;
	junk := _write(1, addr(temp), 4);

	junk := _read( 0, addr(number_workers), 4);
	junk := _write(1, addr(number_workers), 4);

	junk := _read( 0, addr(intervals), 4);
	junk := _write(1, addr(intervals), 4);

	sum := eval(position, number_workers, intervals);

	junk  := _read( 0, addr(total), 8);
	total := total + sum;
	junk  := _write(1, addr(total), 8);
END.
